OpenRA.Game/Exts.cs

```diff
@@ -503,6 +503,26 @@ public static bool IsTraitEnabled<T>(T t)
 		{
 			return IsTraitEnabled(t as object);
 		}
+
+		public static T FirstEnabledTraitOrDefault<T>(this IEnumerable<T> ts)
+		{
+			// PERF: Avoid LINQ.
+			foreach (var t in ts)
+				if (t.IsTraitEnabled())
+					return t;
+
+			return default(T);
+		}
+
+		public static T FirstEnabledTraitOrDefault<T>(this T[] ts)
+		{
+			// PERF: Avoid LINQ.
+			foreach (var t in ts)
```

Comment:
General question: Is the compiler smart enough to generate code to enumerate these by index (`for (var i = 0; i < ts.Length; i++)`, or does it allocate an enumerator object?

Reply:
According to the internet: yes.

Reply:
Yes, the compiler special cases `foreach` on arrays to do exactly that.